File: mobile/android/air/SubModules/WalletContext/src/main/java/org/mytonwallet/app_air/walletcontext/utils/EquatableUtils.kt
package org.mytonwallet.app_air.walletcontext.utils

// Define the WEquatable interface with a method for checking if an item has changed
interface WEquatable<T> {
    fun isSame(comparing: WEquatable<*>): Boolean
    fun isChanged(comparing: WEquatable<*>): Boolean
}

// Define the EquatableChange sealed class with three types: Insert, Delete, and Update
sealed class EquatableChange<T> {
    data class Insert<T>(val item: T) : EquatableChange<T>()
    data class Delete<T>(val item: T) : EquatableChange<T>()
    data class Update<T>(val item: T) : EquatableChange<T>()
}

// Extension function for List<WEquatable> to compute the differences
fun List<WEquatable<*>>.diff(
    newList: List<WEquatable<*>>,
    section: Int = 0
): List<EquatableChange<IndexPath>> {
    val changes = mutableListOf<EquatableChange<IndexPath>>()

    // Find deletions
    for ((index, element) in this.withIndex()) {
        if (newList.find { element.isSame(it) } == null) {
            changes.add(EquatableChange.Delete(IndexPath(index, section)))
        }
    }

    // Find insertions and updates
    for ((index, element) in newList.withIndex()) {
        val oldIndex = indexOfFirst { element.isSame(it) }
        if (oldIndex != -1) {
            if (index != oldIndex || element.isChanged(comparing = this[oldIndex])) {
                // If the element exists in both lists but has moved or changed, consider it as an update
                changes.add(EquatableChange.Update(IndexPath(oldIndex, section)))
            }
        } else {
            // If the element exists only in the new list, consider it as an insertion
            changes.add(EquatableChange.Insert(IndexPath(index, section)))
        }
    }

    return changes
}

// IndexPath to represent the position in section
data class IndexPath(val section: Int, val row: Int)

File: mobile/ios/Air/SubModules/WalletContext/Utils/EquatableUtils.swift
//
//  EquatableUtils.swift
//  WalletContext
//
//  Created by Sina on 3/26/24.
//

import Foundation

public protocol WEquatable: Equatable {
    func isChanged(comparing: Self) -> Bool
}

public enum EquatableChange<T: Equatable> {
    case insert(T)
    case delete(T)
    case update(T)
}

public extension Array where Element: WEquatable {
    func diff(new: [Element], section: Int = 0) -> [EquatableChange<IndexPath>] {
        var changes = [EquatableChange<IndexPath>]()
        
        // Find deletions
        for (index, element) in self.enumerated() {
            if !new.contains(element) {
                changes.append(.delete(IndexPath(row: index, section: section)))
            }
        }
        
        // Find insertions and updates
        for (index, element) in new.enumerated() {
            if let oldIndex = self.firstIndex(of: element) {
                if index != oldIndex || element.isChanged(comparing: self[oldIndex]) {
                    // If the element exists in both arrays but has moved or changed, consider it as an update
                    changes.append(.update(IndexPath(row: oldIndex, section: section)))
                }
            } else {
                // If the element exists only in the new array, consider it as an insertion
                changes.append(.insert(IndexPath(row: index, section: section)))
            }
        }
        
        return changes
    }
    func hasChanged(comparing new: [Element]) -> Bool {
        if self.count != new.count {
            return true
        }
        for (index, element) in enumerated() {
            if (new[index].isChanged(comparing: element)) {
                return true
            }
        }
        return false
    }
}
